<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>aDiff - AGEpy</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "aDiff";
    var mkdocs_page_input_path = "executables/adiff.md";
    var mkdocs_page_url = "/executables/adiff/";
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js"></script>
  <script src="../../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> AGEpy</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../cookbook/">Cookbook</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Modules</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../modules/bed/">bed</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/biom/">biom</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/cytoscape/">cytoscape</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/david/">david</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/fasta/">fasta</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/go/">go</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/gtf/">gtf</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/homology/">homology</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/kegg/">kegg</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/meme/">meme</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/plots/">plots</a>
                </li>
                <li class="">
                    
    <a class="" href="../../modules/sam/">sam</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Executables</span>
    <ul class="subnav">
                <li class=" current">
                    
    <a class="current" href="./">aDiff</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#intro">Intro</a></li>
    

    <li class="toctree-l3"><a href="#examples">Examples</a></li>
    

    <li class="toctree-l3"><a href="#output-files">Output files</a></li>
    

    <li class="toctree-l3"><a href="#help">Help</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../obo2tsv/">obo2tsv</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">AGEpy</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>Executables &raquo;</li>
        
      
    
    <li>aDiff</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h2 id="intro">Intro</h2>
<p><code>aDiff</code> is an annotation tool for differential gene expression results generated by <strong><em>cuffdiff</em></strong> (Trapnell C., <em>Nature Biotechnology</em>, 2012).</p>
<p>It annotates <em>cuffdiff</em> outputs with ensembl gene ids, gene ontology terms and kegg ids.</p>
<p>Additonally it uses <strong><em>DAVID</em></strong>s API (Huang DW, <em>Nature Protoc.</em>, 2009; Huang DW, <em>Nucleic Acids Res.</em>, 2009; Xiaoli J, <em>Bioinformatics</em>, 2012) to perform enrichment analysis.</p>
<p>A <strong><em>Cytoscape</em></strong> (Shannon P, <em>Genome Research</em>, 2003) instance running with the <strong><em>String</em></strong> (Szklarczyk D, <em>Nucleic Acids Res.</em>, 2017) App installed can additionally be plugged in to generate expanded protein-protein interactions.</p>
<p>For a full RNAseq pipeline including <code>aDiff</code> check: <a href="http://bioinformatics.age.mpg.de/presentations-tutorials/presentations/modules/rnaseq-tuxedo-update/#/intro">http://bioinformatics.age.mpg.de/presentations-tutorials/presentations/modules/rnaseq-tuxedo-update/#/intro</a></p>
<h2 id="examples">Examples</h2>
<p>Example of an <code>aDiff</code> call on a <em>c. elegans</em> dataset:</p>
<pre><code>$ aDiff -D -i cuffdiff_output -o adiff_output \
-G references/cel.latest.ensembl.gtf \
-C cuffmerge_output/merged.gtf \
--dataset dmelanogaster_gene_ensembl \
--DAVIDuser &quot;&lt;Registered.Email@david.com&gt;&quot; \
--organismtag CEL \
--cytoscape_host 'localhost' \
--cytoscape_port 1234
</code></pre>

<p>Example of an <code>aDiff</code> call on a <em>d. melanogaster</em> dataset:</p>
<pre><code>$ aDiff -D -i cuffdiff_output -o adiff_output \
-G references/Drosophila_melanogaster.BDGP6.90.gtf \
-C cuffmerge_output/merged.gtf \
--dataset dmelanogaster_gene_ensembl \
--filter flybase_gene_id \
--outputBiotypes 'flybase_gene_id gene_biotype' \
--outputGoterms 'flybase_gene_id go_id name_1006' \
--DAVIDid FLYBASE_GENE_ID \
--DAVIDuser &quot;&lt;Registered.Email@david.com&gt;&quot; \
--organismtag DMEL \
--species 'drosophila melanogaster' \
--cytoscape_host 'localhost' \
--cytoscape_port 1234
</code></pre>

<p>Example of an <code>aDiff</code> call on a <em>mus musculus</em>  dataset:</p>
<pre><code>$ aDiff -i cufdiff_output -o adiff_output \
-G ensembl.mus_musculus.83.original.gtf \
-C cuffmerge_output/merged.gtf \
--TSV \
--dataset mmusculus_gene_ensembl \
-u &quot;&lt;Registered.Email@david.com&gt;&quot; \
--DAVIDid ENSEMBL_GENE_ID \
--host http://dec2015.archive.ensembl.org/biomart \
--organismtag MUS \
--species 'mus musculus' \
--cytoscape_host 'localhost' \
--cytoscape_port 1234
</code></pre>

<p>Example of an <code>aDiff</code> call on a <em>h. sapiens</em>  dataset:</p>
<pre><code>$ aDiff -i cufdiff_output -o adiff_output \
-G ensembl.homo_sapiens.83.original.gtf \
-C cuffmerge_output/merged.gtf \
--TSV \
--dataset hsapiens_gene_ensembl \
-u &quot;&lt;Registered.Email@david.com&gt;&quot; \
--DAVIDid ENSEMBL_GENE_ID \
--host http://dec2015.archive.ensembl.org/biomart \
--organismtag HSA \
--species 'homo sapiens' \
--cytoscape_host 'localhost' \
--cytoscape_port 1234
</code></pre>

<h2 id="output-files">Output files</h2>
<p>Example of the output for the the <em>h. sapiens</em> call above.</p>
<ul>
<li>
<p><strong><code>diff_sig_geneexp.xlsx</code></strong> this file is based on the <em>gene_exp.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it. It contains one sheet for each pairwise comparison filtered to significant values (as defined in <em>cuffdiff</em>).</p>
</li>
<li>
<p><strong><code>diff_sig_iso.xlsx</code></strong> this file is based on the <em>isoform_exp.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it. It contains one sheet for each pairwise comparison filtered to significant values (as defined in <em>cuffdiff</em>).</p>
</li>
<li>
<p><strong><code>diff_sig_prom.xlsx</code></strong> this file is based on the <em>promoters.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it. It contains one sheet for each pairwise comparison filtered to significant values (as defined in <em>cuffdiff</em>).</p>
</li>
<li>
<p><strong><code>diff_sig_splic.xlsx</code></strong> this file is based on the <em>splicing.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it. It contains one sheet for each pairwise comparison filtered to significant values (as defined in <em>cuffdiff</em>).</p>
</li>
<li>
<p><strong><code>diff_sig_cds.xlsx</code></strong> htis file is based on the <em>cds.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it. It contains one sheet for each pairwise comparison filtered to significant values (as defined in <em>cuffdiff</em>).</p>
</li>
<li>
<p><strong><code>geneexp_ALL.tsv</code></strong> this file is based on the <em>gene_exp.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it.</p>
</li>
<li>
<p><strong><code>iso_ALL.tsv</code></strong> this file is based on the <em>isoform_exp.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it.</p>
</li>
<li>
<p><strong><code>prom_ALL.tsv</code></strong> this file is based on the <em>promoters.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it.</p>
</li>
<li>
<p><strong><code>splic_ALL.tsv</code></strong> this file is based on the <em>splicing.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it.</p>
</li>
<li>
<p><strong><code>cds_ALL.tsv</code></strong> this file is based on the <em>cds.diff</em> file output of <em>cuffdiff</em> adding annotation columns to it.</p>
</li>
<li>
<p><strong><code>diff_p.05.xlsx</code></strong> contains a sheet for each of the files above (ie. <em>geneexp_ALL.tsv</em>, <em>iso_ALL.tsv</em>, <em>prom_ALL.tsv</em>, <em>splic_ALL.tsv</em>, <em>cds_ALL.tsv</em> ) subset to p values bellow 0.05.</p>
</li>
<li>
<p><strong><code>KEGG_PATHWAY_diff_sig_geneexp.xlsx</code></strong> this file is based on the <em>gene_exp.diff</em> file output of <em>cuffdiff</em>. It generates a result sheet for each pairwise comparison. It reports DAVID enrichment results for KEGG using genes labeled as significant by <em>cuffdiff</em>.</p>
</li>
<li>
<p><strong><code>GOTERM_BP_FAT_diff_sig_splic.xlsx</code></strong> this is file is based on the <em>splicing.diff</em> file output of <em>cuffdiff</em>.  It generates a result sheet for each pairwise comparison. It reports DAVID enrichment results for Gene Ontology Biological Process (GOTERM BP) using genes labeled as significant by <em>cuffdiff</em>.</p>
</li>
<li>
<p><strong><code>OMIM_DISEASE_diff_sig_geneexp.xlsx</code></strong> this file is based on the <em>gene_exp.diff</em> file output of <em>cuffdiff</em>. It generates a result sheet for each pairwise comparison. It reports DAVID enrichment results for OMIM DISEASE using genes labeled as significant by <em>cuffdiff</em>.</p>
</li>
</ul>
<p>More information on the standard ouput columns of <em>cuffdiff</em> can be found <a href="http://cole-trapnell-lab.github.io/cufflinks/cuffdiff/index.html">here</a>.</p>
<h2 id="help">Help</h2>
<pre><code>$ aDiff --help

usage: aDiff [-h] [-D] [-i INPUTFOLDER] [-o OUTPUTFOLDER] [-G ORIGINALGTF]
             [-C CUFFCOMPAREGTF] [-f INPUTFILES] [-s SHORTOUTPUTNAME]
             [--sigOnly] [--TSV] [--TSVall] [--description] [--listMarts]
             [--mart MART] [--listDatasets] [--dataset DATASET]
             [--listFilters] [--filter FILTER] [--listAttributes]
             [--outputBiotypes OUTPUTBIOTYPES] [--outputGoterms OUTPUTGOTERMS]
             [--KEGG] [--listKEGGorganisms] [--KEGGorg KEGGORG] [--findKEGGdb]
             [--KEGGdb KEGGDB] [--DAVIDid DAVIDID] [--DAVIDcat DAVIDCAT]
             [-u DAVIDUSER] [--host HOST] [--organismtag {DMEL,CEL,MUS,HSA}]
             [--species SPECIES] [--limit LIMIT] [--cuttoff CUTTOFF]
             [--taxon TAXON] [--cytoscape_host CYTOSCAPE_HOST]
             [--cytoscape_port CYTOSCAPE_PORT]

optional arguments:
  -h, --help            show this help message and exit
  -D, --DAVID           Use this flag to perform DAVID GO enrichment analysis
                        (default: False)
  -i INPUTFOLDER, --inputFolder INPUTFOLDER
                        Cuffdiff output folder (default: None)
  -o OUTPUTFOLDER, --outputFolder OUTPUTFOLDER
                        Output folder (default: None)
  -G ORIGINALGTF, --originalGTF ORIGINALGTF
                        Original/downloaded GTF (default: None)
  -C CUFFCOMPAREGTF, --cuffcompareGTF CUFFCOMPAREGTF
                        Merged cuffcompared GTF (default: None)
  -f INPUTFILES, --inputFiles INPUTFILES
                        Implies -s. Use this option to select which *.diff
                        files you wish to analyse.'. (default: gene_exp.diff
                        promoters.diff splicing.diff cds.diff
                        isoform_exp.diff)
  -s SHORTOUTPUTNAME, --shortOutputName SHORTOUTPUTNAME
                        Use this option to select a short outpput name for
                        each *.diff file used in '-f'. No '.' (dots) allowed.
                        (default: geneexp prom splic cds iso)
  --sigOnly             Only create report tables for cuffdiff-labeled
                        significantly changed genes (default: False)
  --TSV                 For p values &gt; = 0.05 write tables as tab separated
                        values (default: False)
  --TSVall              Save p &lt; 0.05 save tables as tab separated values in a
                        folder called TSV (default: False)
  --description         Get a description of what this script does. (default:
                        False)
  --listMarts           List biomaRt Marts (default: False)
  --mart MART           Your mart of choice. (default: ENSEMBL_MART_ENSEMBL)
  --listDatasets        List datasets for your mart (default: False)
  --dataset DATASET     Dataset of your choice. (default:
                        celegans_gene_ensembl)
  --listFilters         List available filters (default: False)
  --filter FILTER       Filter to use to identify your genes. (default:
                        ensembl_gene_id)
  --listAttributes      List available attributes for your dataset. (default:
                        False)
  --outputBiotypes OUTPUTBIOTYPES
                        Outputs/attributes for your biotypes data. Order has
                        to be kept, ie. first IDs then biotype. (default:
                        ensembl_gene_id gene_biotype)
  --outputGoterms OUTPUTGOTERMS
                        Outputs/attributes for your goterms data. Order has to
                        be kept, ie. 1st gene_id, then go_id, then
                        go_term_name (default: ensembl_gene_id go_id
                        name_1006)
  --KEGG                Add KEGG annotations (default: False)
  --listKEGGorganisms   List KEGG organisms. (default: False)
  --KEGGorg KEGGORG     KEGG organism. (default: cel)
  --findKEGGdb          KEGG has DB identifier for each linked DB. Use this
                        function to find the label of your DB, eg: 'ensembl-
                        hsa', 'FlyBase'. This option requires --originalGTF
                        and --KEGGorg (default: False)
  --KEGGdb KEGGDB       KEGG database linked to your ensembl organism.
                        (default: EnsemblGenomes-Gn)
  --DAVIDid DAVIDID     DAVID's id for your dataset. List of ids available in
                        http://david.abcc.ncifcrf.gov/content.jsp?file=DAVID_A
                        PI.html#input_list (default: WORMBASE_GENE_ID)
  --DAVIDcat DAVIDCAT   DAVID's categories you wish to analyse. List of
                        available categories in https://david.ncifcrf.gov/cont
                        ent.jsp?file=DAVID_API.html#approved_list. (default: G
                        OTERM_BP_FAT,GOTERM_CC_FAT,GOTERM_MF_FAT,KEGG_PATHWAY,
                        PFAM,PROSITE,GENETIC_ASSOCIATION_DB_DISEASE,OMIM_DISEA
                        SE)
  -u DAVIDUSER, --DAVIDuser DAVIDUSER
                        Your DAVID's user id. example: 'John.Doe@age.mpg.de'
                        (default: None)
  --host HOST           Ensembl host. Check http://www.ensembl.org/info/websit
                        e/archives/index.html for older releases. (default:
                        http://www.ensembl.org/biomart)
  --organismtag {DMEL,CEL,MUS,HSA}
                        Organism tag. (default: None)
  --species SPECIES     Species for string app query. eg. 'caenorhabditis
                        elegans', 'drosophila melanogaster', 'mus musculus',
                        'homo sapiens'. Default='caenorhabditis elegans'
                        (default: caenorhabditis elegans)
  --limit LIMIT         Limit for string app query. Number of extra genes to
                        recover. If None, limit=N(query_genes)*.25 (default:
                        None)
  --cuttoff CUTTOFF     Confidence cuttoff for sting app query. Default=0.4
                        (default: 0.4)
  --taxon TAXON         Taxon id for string app query. For the species shown
                        above, taxon id will be automatically identified.
                        (default: None)
  --cytoscape_host CYTOSCAPE_HOST
                        Host address for cytoscape. (default: None)
  --cytoscape_port CYTOSCAPE_PORT
                        Cytoscape port. (default: None)
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../obo2tsv/" class="btn btn-neutral float-right" title="obo2tsv">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../../modules/sam/" class="btn btn-neutral" title="sam"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../../modules/sam/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../obo2tsv/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/require.js"></script>
      <script src="../../search/search.js"></script>

</body>
</html>
