<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>gtf - AGEpy</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "gtf";
    var mkdocs_page_input_path = "modules/gtf.md";
    var mkdocs_page_url = "/modules/gtf/";
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js"></script>
  <script src="../../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> AGEpy</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../cookbook/">Cookbook</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Modules</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../bed/">bed</a>
                </li>
                <li class="">
                    
    <a class="" href="../biom/">biom</a>
                </li>
                <li class="">
                    
    <a class="" href="../cytoscape/">cytoscape</a>
                </li>
                <li class="">
                    
    <a class="" href="../david/">david</a>
                </li>
                <li class="">
                    
    <a class="" href="../fasta/">fasta</a>
                </li>
                <li class="">
                    
    <a class="" href="../go/">go</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">gtf</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#readgtf">readGTF</a></li>
    

    <li class="toctree-l3"><a href="#retrieve_gtf_field">retrieve_GTF_field</a></li>
    

    <li class="toctree-l3"><a href="#attributesgtf">attributesGTF</a></li>
    

    <li class="toctree-l3"><a href="#parsegtf">parseGTF</a></li>
    

    <li class="toctree-l3"><a href="#writegtf">writeGTF</a></li>
    

    <li class="toctree-l3"><a href="#mapgenototrans">MAPGenoToTrans</a></li>
    

    <li class="toctree-l3"><a href="#gtftobed">GTFtoBED</a></li>
    

    <li class="toctree-l3"><a href="#gettransposition">GetTransPosition</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../homology/">homology</a>
                </li>
                <li class="">
                    
    <a class="" href="../kegg/">kegg</a>
                </li>
                <li class="">
                    
    <a class="" href="../meme/">meme</a>
                </li>
                <li class="">
                    
    <a class="" href="../plots/">plots</a>
                </li>
                <li class="">
                    
    <a class="" href="../sam/">sam</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Executables</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../executables/adiff/">aDiff</a>
                </li>
                <li class="">
                    
    <a class="" href="../../executables/obo2tsv/">obo2tsv</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">AGEpy</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>Modules &raquo;</li>
        
      
    
    <li>gtf</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h2 id="readgtf"><strong><em>readGTF</em></strong></h2>
<p>Reads a GTF file and labels the respective columns in agreement with GTF file standards:
'seqname','source','feature','start','end','score','strand','frame','attribute'.</p>
<p><strong><code>readGTF(infile)</code></strong></p>
<ul>
<li><strong><code>infile</code></strong> /path/to/file.gtf</li>
<li><strong><code>returns</code></strong> a Pandas dataframe of the respective GTF</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; GTF=age.readGTF(&quot;gencode.v24.primary_assembly.annotation.gtf&quot;)
&gt;&gt;&gt; print GTF.head()

seqname  source     feature  start    end score strand frame  \
0    chr1  HAVANA        gene  11869  14409     .      +     .   
1    chr1  HAVANA  transcript  11869  14409     .      +     .   
2    chr1  HAVANA        exon  11869  12227     .      +     .   
3    chr1  HAVANA        exon  12613  12721     .      +     .   
4    chr1  HAVANA        exon  13221  14409     .      +     .   

                                         attribute  
0  gene_id &quot;ENSG00000223972.5&quot;; gene_type &quot;transc...&quot;  
1  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;
</code></pre>

<hr />
<h2 id="retrieve_gtf_field"><strong><em>retrieve_GTF_field</em></strong></h2>
<p>Returns a field of choice from the attribute column of the GTF.</p>
<p><strong><code>retrieve_GTF_field(field,gtf)</code></strong></p>
<ul>
<li><strong><code>field</code></strong> field to be retrieved</li>
<li><strong><code>returns</code></strong> a Pandas dataframe with one column containing the field of choice</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; GTF=age.readGTF(&quot;/gencode.v24.primary_assembly.annotation.gtf&quot;)
&gt;&gt;&gt; print GTF.head()

seqname  source     feature  start    end score strand frame  \
0    chr1  HAVANA        gene  11869  14409     .      +     .   
1    chr1  HAVANA  transcript  11869  14409     .      +     .   
2    chr1  HAVANA        exon  11869  12227     .      +     .   
3    chr1  HAVANA        exon  12613  12721     .      +     .   
4    chr1  HAVANA        exon  13221  14409     .      +     .   

                                         attribute  
0  gene_id &quot;ENSG00000223972.5&quot;; gene_type &quot;transc...&quot;  
1  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;

&gt;&gt;&gt; GTF[&quot;gene_id&quot;]=age.retrieve_GTF_field(&quot;gene_id&quot;,GTF)
&gt;&gt;&gt; print GTF.head()

seqname  source     feature  start    end score strand frame  \
0    chr1  HAVANA        gene  11869  14409     .      +     .   
1    chr1  HAVANA  transcript  11869  14409     .      +     .   
2    chr1  HAVANA        exon  11869  12227     .      +     .   
3    chr1  HAVANA        exon  12613  12721     .      +     .   
4    chr1  HAVANA        exon  13221  14409     .      +     .   

                                         attribute            gene_id  
0  gene_id &quot;ENSG00000223972.5&quot;; gene_type &quot;transc...&quot;  ENSG00000223972.5  
1  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5  
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5  
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5  
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5
</code></pre>

<hr />
<h2 id="attributesgtf"><strong><em>attributesGTF</em></strong></h2>
<p>List the type of attributes in a the attribute section of a GTF file</p>
<p><strong><code>attributesGTF(inGTF)</code></strong></p>
<ul>
<li><strong><code>inGTF</code></strong> GTF dataframe to be analysed</li>
<li><strong><code>returns</code></strong> a list of attributes present in the attribute section</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; attributes=age.attributesGTF(GTF)
&gt;&gt;&gt; print attributes

['gene_status', 'havana_gene', 'transcript_support_level', 'level', 'transcript_type', 'tag', 'protein_id', 'gene_id', 'exon_id', 'transcript_id', 'exon_number', 'ont', 'havana_transcript', 'ccdsid', 'transcript_name', 'gene_type', 'transcript_status', 'gene_name']
</code></pre>

<hr />
<h2 id="parsegtf"><strong><em>parseGTF</em></strong></h2>
<p>Reads an extracts all attributes in the attributes section of a GTF and constructs a new dataframe wiht one collumn per attribute instead of the attributes column.</p>
<p><strong><code>parseGTF(inGTF)</code></strong></p>
<ul>
<li><strong><code>inGTF</code></strong> GTF dataframe to be parsed</li>
<li><strong><code>returns</code></strong> a dataframe of the orignal input GTF with attributes parsed</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; GTF=age.readGTF(&quot;gencode.v24.primary_assembly.annotation.gtf&quot;)
&gt;&gt;&gt; print GTF.head()

seqname  source     feature  start    end score strand frame  \
0    chr1  HAVANA        gene  11869  14409     .      +     .   
1    chr1  HAVANA  transcript  11869  14409     .      +     .   
2    chr1  HAVANA        exon  11869  12227     .      +     .   
3    chr1  HAVANA        exon  12613  12721     .      +     .   
4    chr1  HAVANA        exon  13221  14409     .      +     .   

                                         attribute  
0  gene_id &quot;ENSG00000223972.5&quot;; gene_type &quot;transc...&quot;  
1  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  

&gt;&gt;&gt; GTFpa=age.parseGTF(GTF)
&gt;&gt;&gt; print GTFpa.head()

seqname  source     feature  start    end score strand frame gene_status  \
0    chr1  HAVANA        gene  11869  14409     .      +     .       KNOWN   
1    chr1  HAVANA  transcript  11869  14409     .      +     .       KNOWN   
2    chr1  HAVANA        exon  11869  12227     .      +     .       KNOWN   
3    chr1  HAVANA        exon  12613  12721     .      +     .       KNOWN   
4    chr1  HAVANA        exon  13221  14409     .      +     .       KNOWN   

            havana_gene    ...               exon_id      transcript_id  \
0  OTTHUMG00000000961.2    ...                   NaN                NaN   
1  OTTHUMG00000000961.2    ...                   NaN  ENST00000456328.2   
2  OTTHUMG00000000961.2    ...     ENSE00002234944.1  ENST00000456328.2   
3  OTTHUMG00000000961.2    ...     ENSE00003582793.1  ENST00000456328.2   
4  OTTHUMG00000000961.2    ...     ENSE00002312635.1  ENST00000456328.2   

  exon_number  ont     havana_transcript ccdsid transcript_name  \
0         NaN  NaN                   NaN    NaN             NaN   
1         NaN  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   
2           1  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   
3           2  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   
4           3  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   

                            gene_type transcript_status gene_name  
0  transcribed_unprocessed_pseudogene               NaN   DDX11L1  
1  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1  
2  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1  
3  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1  
4  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1
</code></pre>

<hr />
<h2 id="writegtf"><strong><em>writeGTF</em></strong></h2>
<p>Write a GTF dataframe into a file.</p>
<p><strong><code>writeGTF(inGTF,file_path)</code></strong></p>
<ul>
<li><strong><code>inGTF</code></strong> GTF dataframe to be written. It should either have 9 columns with the last one being the "attributes" section or more than 9 columns where all columns after the 8th will be colapsed into one.</li>
<li><strong><code>file_path</code></strong> /path/to/the/file.gtf</li>
<li><strong><code>returns</code></strong> nothing</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; writeGTF(GTFpa,&quot;/path/to/new/file.gtf&quot;)
</code></pre>

<hr />
<h2 id="mapgenototrans"><strong><em>MAPGenoToTrans</em></strong></h2>
<p>Gets all positions of all bases in an exon.</p>
<p><strong><code>MAPGenoToTrans(parsedGTF,feature)</code></strong></p>
<ul>
<li><strong><code>df</code></strong> a Pandas dataframe with 'start','end', and 'strand' information for each entry. df must contain ['seqname','feature','start','end','strand','frame','gene_id',  'transcript_id','exon_id','exon_number']</li>
<li><strong><code>feature</code></strong> feature upon wich to generate the map, eg. 'exon' or 'transcript'</li>
<li><strong><code>returns</code></strong> a dictionary with a string with the comma separated positions of all bases in the exon</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; print GTF.head()

seqname  source     feature  start    end score strand frame  \
0    chr1  HAVANA        gene  11869  14409     .      +     .   
1    chr1  HAVANA  transcript  11869  14409     .      +     .   
2    chr1  HAVANA        exon  11869  12227     .      +     .   
3    chr1  HAVANA        exon  12613  12721     .      +     .   
4    chr1  HAVANA        exon  13221  14409     .      +     .   

                                         attribute            gene_id  \
0  gene_id &quot;ENSG00000223972.5&quot;; gene_type &quot;transc...&quot;  ENSG00000223972.5   
1  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   

     transcript_id            exon_id exon_number  
0                NaN                NaN         NaN  
1  ENST00000456328.2                NaN         NaN  
2  ENST00000456328.2  ENSE00002234944.1           1  
3  ENST00000456328.2  ENSE00003582793.1           2  
4  ENST00000456328.2  ENSE00002312635.1           3

&gt;&gt;&gt; GtoT=age.MAPGenoToTrans(GTF,&quot;exon&quot;)
&gt;&gt;&gt; print GtoT

{ENST23923910:'234,235,236,1021,..'}
</code></pre>

<hr />
<h2 id="gtftobed"><strong><em>GTFtoBED</em></strong></h2>
<p>Transform a GTF dataframe into a bed dataframe</p>
<p><strong><code>GTFtoBED(inGTF,name)</code></strong></p>
<ul>
<li><strong><code>inGTF</code></strong> GTF dataframe for transformation</li>
<li><strong><code>name</code></strong> field of the GTF data frame to be use for the bed 'name' positon</li>
<li><strong><code>returns</code></strong> a bed dataframe with the corresponding bed fiels: 'chrom','chromStart','chromEnd','name','score','strand'</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; bed = age.GTFtoBED(GTF, &quot;gene_id&quot;)
</code></pre>

<hr />
<h2 id="gettransposition"><strong><em>GetTransPosition</em></strong></h2>
<p>Maps a genome position to transcript positon.</p>
<p><strong><code>GetTransPosition(df, field, dic, refCol="transcript_id")</code></strong></p>
<ul>
<li><strong><code>df</code></strong> a Pandas dataframe</li>
<li><strong><code>field</code></strong> the head of the column containing the genomic position</li>
<li><strong><code>dic</code></strong> a dictionary containing for each transcript the respective bases eg. {ENST23923910:'234,235,236,1021,..'}. See <em>MAPGenoToTrans</em>.</li>
<li><strong><code>refCol</code></strong> header of the reference column with IDs, eg. 'transcript_id'</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; print GTF_.head()

seqname  source feature  start    end score strand frame  \
2    chr1  HAVANA    exon  11869  12227     .      +     .   
3    chr1  HAVANA    exon  12613  12721     .      +     .   
4    chr1  HAVANA    exon  13221  14409     .      +     .   
6    chr1  HAVANA    exon  12010  12057     .      +     .   
7    chr1  HAVANA    exon  12179  12227     .      +     .   

                                         attribute            gene_id  \
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
6  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
7  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   

     transcript_id            exon_id exon_number  target  
2  ENST00000456328.2  ENSE00002234944.1           1   12000  
3  ENST00000456328.2  ENSE00003582793.1           2   12617  
4  ENST00000456328.2  ENSE00002312635.1           3   14000  
6  ENST00000450305.2  ENSE00001948541.1           1   12040  
7  ENST00000450305.2  ENSE00001671638.2           2   12210  

&gt;&gt;&gt; GTF_[&quot;transcript target&quot;]=GTF_.apply(age.GetTransPosition, \
                                     args=(&quot;target&quot;,GtoT),axis=1)
&gt;&gt;&gt; print GTF_.head()

seqname  source feature  start    end score strand frame  \
2    chr1  HAVANA    exon  11869  12227     .      +     .   
3    chr1  HAVANA    exon  12613  12721     .      +     .   
4    chr1  HAVANA    exon  13221  14409     .      +     .   
6    chr1  HAVANA    exon  12010  12057     .      +     .   
7    chr1  HAVANA    exon  12179  12227     .      +     .   

                                         attribute            gene_id  \
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
6  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   
7  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  ENSG00000223972.5   

     transcript_id            exon_id exon_number  target  transcript target  
2  ENST00000456328.2  ENSE00002234944.1           1   12000                132  
3  ENST00000456328.2  ENSE00003582793.1           2   12617                364  
4  ENST00000456328.2  ENSE00002312635.1           3   14000               1248  
6  ENST00000450305.2  ENSE00001948541.1           1   12040                 31  
7  ENST00000450305.2  ENSE00001671638.2           2   12210                 80  
</code></pre>

<hr />
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../homology/" class="btn btn-neutral float-right" title="homology">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../go/" class="btn btn-neutral" title="go"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../go/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../homology/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/require.js"></script>
      <script src="../../search/search.js"></script>

</body>
</html>
