<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>plots - AGEpy</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "plots";
    var mkdocs_page_input_path = "modules/plots.md";
    var mkdocs_page_url = "/modules/plots/";
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js"></script>
  <script src="../../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> AGEpy</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../cookbook/">Cookbook</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Modules</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../bed/">bed</a>
                </li>
                <li class="">
                    
    <a class="" href="../biom/">biom</a>
                </li>
                <li class="">
                    
    <a class="" href="../cytoscape/">cytoscape</a>
                </li>
                <li class="">
                    
    <a class="" href="../david/">david</a>
                </li>
                <li class="">
                    
    <a class="" href="../fasta/">fasta</a>
                </li>
                <li class="">
                    
    <a class="" href="../go/">go</a>
                </li>
                <li class="">
                    
    <a class="" href="../gtf/">gtf</a>
                </li>
                <li class="">
                    
    <a class="" href="../homology/">homology</a>
                </li>
                <li class="">
                    
    <a class="" href="../kegg/">kegg</a>
                </li>
                <li class="">
                    
    <a class="" href="../meme/">meme</a>
                </li>
                <li class=" current">
                    
    <a class="current" href="./">plots</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#cellplot">CellPlot</a></li>
    

    <li class="toctree-l3"><a href="#symplot">SymPlot</a></li>
    

    <li class="toctree-l3"><a href="#ma">MA</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../sam/">sam</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Executables</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../executables/adiff/">aDiff</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">AGEpy</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>Modules &raquo;</li>
        
      
    
    <li>plots</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h2 id="cellplot"><strong><em>CellPlot</em></strong></h2>
<p>Python implementation of the CellPlot from the CellPlot package for R.
-inf or inf enrichments will come out as min found float or max found float, respectively.</p>
<ul>
<li><strong><code>df</code></strong> pandas dataframe with the following columns - 'Enrichment', 'Term', and 'log2fc'.
           For log2fc each cell must contain a comma separated string with the log2fc for the genes enriched in the respective term.
           eg. '-inf,-1,2,3.4,3.66,inf'</li>
<li><strong><code>output_file</code></strong> prefix for an output file. If given it will create output_file.CellPlot.svg and output_file.CellPlot.png</li>
<li><strong><code>gene_expression</code></strong> label for the color gradiant bar.</li>
<li><strong><code>figure_title</code></strong> Figure title.</li>
<li><strong><code>pvalCol</code></strong> name of the column containing the p values to determine if the terms should be marked as NS - not significant, use None for no marking</li>
<li><strong><code>lowerLimit</code></strong> lower limit for the heatmap bar (default is the 0.1 percentile)</li>
<li><strong><code>upperLimit</code></strong> upper limit for the heatmap bar (default is the 0.9 percentile)</li>
<li><strong><code>colorBarType</code></strong> type of heatmap, 'Spectral' is default, alternative eg. 'seismic'</li>
<li><strong><code>returns</code></strong> a matplotlib figure</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; print df.head()

Term  Annotated  Enrichment  \
0          GO:0008544~epidermis development         38    4.006021   
1               GO:0043588~skin development         33    4.359840   
2         GO:0045087~innate immune response         61    2.385984   
3               GO:0006952~defense response         90    1.913315   
4  GO:0009605~response to external stimulus        113    1.736641   

ease                                             log2fc  
0  1.193931e-12  1.13845,0.771811,0.926561,0.578588,-0.694105,1...  
1  4.757460e-12  1.13845,0.926561,-0.694105,1.48945,0.94486,-1....  
2  5.609421e-10  -1.91507,-0.630414,-1.87466,-0.898252,0.458041...  
3  2.238959e-09  -0.538926,0.667335,-1.91507,-0.630414,-1.87466...  
4  3.051460e-09  0.667335,-1.91507,-0.630414,1.46227,0.755911,-...  

&gt;&gt;&gt; cellplot=age.CellPlot(df[:20], &quot;cellplot&quot;,  &quot;log2(mt/wt)&quot;, &quot;mutant 1&quot;, \
pvalCol=&quot;ease&quot;, colorBarType=&quot;bwr&quot;, lowerLimit=-1.25,upperLimit=1.25)
</code></pre>

<p><img alt="cellpot" src="../cellplot.CellPlot.png" /></p>
<hr />
<h2 id="symplot"><strong><em>SymPlot</em></strong></h2>
<p>Python implementation of the SymPlot from the CellPlot package for R.
-inf or inf enrichments will come out as min found float or max found float, respectively.</p>
<ul>
<li><strong><code>df</code></strong> pandas dataframe with the following columns - 'Enrichment', 'Significant', 'Annotated', 'Term', and 'log2fc'. 'Annotated'i stands for number of genes annotated with the respective GO term. As reported in DAVID by listHits. For log2fc each cell must contain a comma separated string with the log2fc for the genes enriched in the respective term. eg. '-inf,-1,2,3.4,3.66,inf'</li>
<li><strong><code>output_file</code></strong> prefix for an output file. If given it witll create output_file.SymPlot.svg and output_file.SymPlot.png</li>
<li><strong><code>figure_title</code></strong> Figure title.</li>
<li><strong><code>pvalCol</code></strong> name of the column containing the p values to determine if the terms should be marked as NS - not significant, use None for no marking</li>
<li><strong><code>returns</code></strong> a matplotlib figure</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; symplot=age.SymPlot(df[:20],&quot;symplot&quot;, &quot;mutant 1&quot;,pvalCol=&quot;ease&quot;)
</code></pre>

<p><img alt="sympot" src="../symplot.SymPlot.png" /></p>
<hr />
<h2 id="ma"><strong><em>MA</em></strong></h2>
<p>Plots an MA like plot.</p>
<ul>
<li><strong><code>df</code></strong> dataframe output of GetData()</li>
<li><strong><code>title</code></strong> plot title, 'Genes' or 'Transcripts'</li>
<li><strong><code>figName</code></strong> /path/to/saved/figure/prefix</li>
<li><strong><code>c</code></strong> pair of samples to be plotted in list format</li>
<li><strong><code>daType</code></strong> data type, ie. 'counts' or 'FPKM'</li>
<li><strong><code>nbins</code></strong> number of bins on normalized intensities to fit the splines</li>
<li><strong><code>per</code></strong> log2(fold change) percentil to which the splines will be fitted</li>
<li><strong><code>deg</code></strong> degress of freedom used to fit the splines</li>
<li><strong><code>eq</code></strong> if true assumes for each bin that the lower and upper values are equally distant to 0, taking the smaller distance for both</li>
<li><strong><code>splines</code></strong> plot splines, default=True</li>
<li><strong><code>spec</code></strong> list of ids to be highlighted</li>
<li><strong><code>Targets</code></strong> list of ids that will be highlighted if outside of the fitted splines</li>
<li><strong><code>ylim</code></strong> a list of limits to apply on the y-axis of the plot</li>
<li><strong><code>sizeRed</code></strong> size of the highlight marker</li>
<li><strong><code>returns df_</code></strong> a Pandas dataframe similar to the GetData() output with normalized intensities and spline outbounds rows marked as 1.</li>
<li><strong><code>returns red</code></strong> list of ids that are highlighted</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age
&gt;&gt;&gt; print df.head()

gene_id                    gene  wt0  wt20  log2(wt20/wt0)  \
0  ENSG00000223972                 DDX11L1  0.0   0.0             NaN   
1  ENSG00000243485  MIR1302-2,RP11-34P13.3  0.0   0.0             NaN   
2  ENSG00000274890  MIR1302-2,RP11-34P13.3  0.0   0.0             NaN   
3  ENSG00000268020                  OR4G4P  0.0   0.0             NaN   
4  ENSG00000240361                 OR4G11P  0.0   0.0             NaN   

p_value  q_value significant  
0      1.0      1.0          no  
1      1.0      1.0          no  
2      1.0      1.0          no  
3      1.0      1.0          no  
4      1.0      1.0          no  

&gt;&gt;&gt; madf1,sig1=age.MA(dge_, 'Genes',&quot;MA1&quot;,[&quot;wt0&quot;,&quot;wt20&quot;], daType=&quot;FPKM&quot;)
</code></pre>

<p><img alt="ma1" src="../MA1.png" /></p>
<pre><code class="python">&gt;&gt;&gt; sigGenes=df[df[&quot;significant&quot;==&quot;yes&quot;]][&quot;gene_id&quot;].tolist()
&gt;&gt;&gt; madf2,sig2=age.MA(dge_, 'Genes',&quot;MA2&quot;, [&quot;wt0&quot;,&quot;wt20&quot;], splines=False, daType=&quot;FPKM&quot;,spec=sigGenes)
</code></pre>

<p><img alt="ma2" src="../MA2.png" /></p>
<pre><code class="python">&gt;&gt;&gt; madf3,sig3=age.MA(dge_, 'Genes',&quot;MA3&quot;, [&quot;wt0&quot;,&quot;wt20&quot;], splines=True, daType=&quot;FPKM&quot;,Targets=sigGenes)
</code></pre>

<p><img alt="ma3" src="../MA3.png" /></p>
<hr />
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../sam/" class="btn btn-neutral float-right" title="sam">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../meme/" class="btn btn-neutral" title="meme"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../meme/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../sam/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/require.js"></script>
      <script src="../../search/search.js"></script>

</body>
</html>
