<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>bed - AGEpy</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="../../css/highlight.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "bed";
    var mkdocs_page_input_path = "modules/bed.md";
    var mkdocs_page_url = "/modules/bed/";
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js"></script>
  <script src="../../js/modernizr-2.8.3.min.js"></script>
  <script type="text/javascript" src="../../js/highlight.pack.js"></script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> AGEpy</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <a class="" href="../../cookbook/">Cookbook</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Modules</span>
    <ul class="subnav">
                <li class=" current">
                    
    <a class="current" href="./">bed</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#getbednarrowpeakgz">GetBEDnarrowPeakgz</a></li>
    

    <li class="toctree-l3"><a href="#writebed">writeBED</a></li>
    

    <li class="toctree-l3"><a href="#dftobedtool">dfTObedtool</a></li>
    

    <li class="toctree-l3"><a href="#getpeaksexons">GetPeaksExons</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../biom/">biom</a>
                </li>
                <li class="">
                    
    <a class="" href="../cytoscape/">cytoscape</a>
                </li>
                <li class="">
                    
    <a class="" href="../david/">david</a>
                </li>
                <li class="">
                    
    <a class="" href="../fasta/">fasta</a>
                </li>
                <li class="">
                    
    <a class="" href="../go/">go</a>
                </li>
                <li class="">
                    
    <a class="" href="../gtf/">gtf</a>
                </li>
                <li class="">
                    
    <a class="" href="../homology/">homology</a>
                </li>
                <li class="">
                    
    <a class="" href="../kegg/">kegg</a>
                </li>
                <li class="">
                    
    <a class="" href="../meme/">meme</a>
                </li>
                <li class="">
                    
    <a class="" href="../plots/">plots</a>
                </li>
                <li class="">
                    
    <a class="" href="../sam/">sam</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Executables</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../executables/adiff/">aDiff</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">AGEpy</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>Modules &raquo;</li>
        
      
    
    <li>bed</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h2 id="getbednarrowpeakgz"><strong><em>GetBEDnarrowPeakgz</em></strong></h2>
<p>Reads a gz compressed BED narrow peak file from a web address or local file and returns a pandas dataframe.</p>
<ul>
<li>
<p><strong><code>URL_or_PATH_TO_file</code></strong> source of input bed. Either a web link or a path to a local file.</p>
</li>
<li>
<p><strong><code>returns</code></strong> a pandas dataframe of the inpud bed.</p>
</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age

&gt;&gt;&gt; eCLIP_1_bednarrowPeak=&quot;https://www.encodeproject.org/files/ENCFF066PCT/@@download/ENCFF066PCT.bed.gz&quot;
&gt;&gt;&gt; bed=age.GetBEDnarrowPeakgz(eCLIP_1_bednarrowPeak)
&gt;&gt;&gt; print bed.head()

chrom chromStart   chromEnd    name score strand       signalValue  \
0  chr7  139371278  139371296  Peak_0  1000      +  5.09062636514014   
1  chr7  139371257  139371278  Peak_1  1000      +   5.0840236303159   
2  chr7  155781335  155781431  Peak_2  1000      +  3.70481328524336   
3  chr7   87156569   87156676  Peak_3  1000      +  3.95023151551588   
4  chr7  105073472  105073521  Peak_4  1000      +  4.14556204062503   

     -log10(pValue) -log10(qvalue) peak  
0  48.9834262537309             -1   -1  
1  48.7463712698062             -1   -1  
2  42.6519289009201             -1   -1  
3  37.7848384917051             -1   -1  
4  34.0756845242392             -1   -1
</code></pre>

<hr />
<h2 id="writebed"><strong><em>writeBED</em></strong></h2>
<p>Writes a bed dataframe into a bed file.</p>
<ul>
<li><strong><code>inBED</code></strong> a pandas dataframe with the contents of the bed file to be written.</li>
<li>
<p><strong><code>file_path</code></strong> path to target file.</p>
</li>
<li>
<p><strong><code>returns</code></strong> nothing.</p>
</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age

&gt;&gt;&gt; print bed.head()

chrom chromStart   chromEnd    name score strand       signalValue  \
0  chr7  139371278  139371296  Peak_0  1000      +  5.09062636514014   
1  chr7  139371257  139371278  Peak_1  1000      +   5.0840236303159   
2  chr7  155781335  155781431  Peak_2  1000      +  3.70481328524336   
3  chr7   87156569   87156676  Peak_3  1000      +  3.95023151551588   
4  chr7  105073472  105073521  Peak_4  1000      +  4.14556204062503   

     -log10(pValue) -log10(qvalue) peak  
0  48.9834262537309             -1   -1  
1  48.7463712698062             -1   -1  
2  42.6519289009201             -1   -1  
3  37.7848384917051             -1   -1  
4  34.0756845242392             -1   -1

&gt;&gt;&gt; age.writeBED(bed,&quot;/path/to/file.bed&quot;)
</code></pre>

<hr />
<h2 id="dftobedtool"><strong><em>dfTObedtool</em></strong></h2>
<p>Transforms a pandas dataframe into a bedtool. Requires <code>bedtools</code> to be in your <code>path</code>.</p>
<ul>
<li><strong><code>df</code></strong> a pandas dataframe.</li>
<li><strong><code>returns</code></strong> a bedtool.</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age

&gt;&gt;&gt; print bed.head()

chrom chromStart   chromEnd    name score strand       signalValue  \
0  chr7  139371278  139371296  Peak_0  1000      +  5.09062636514014   
1  chr7  139371257  139371278  Peak_1  1000      +   5.0840236303159   
2  chr7  155781335  155781431  Peak_2  1000      +  3.70481328524336   
3  chr7   87156569   87156676  Peak_3  1000      +  3.95023151551588   
4  chr7  105073472  105073521  Peak_4  1000      +  4.14556204062503   

     -log10(pValue) -log10(qvalue) peak  
0  48.9834262537309             -1   -1  
1  48.7463712698062             -1   -1  
2  42.6519289009201             -1   -1  
3  37.7848384917051             -1   -1  
4  34.0756845242392             -1   -1

&gt;&gt;&gt; bedtool=age.dfTObedtool(bed)
&gt;&gt;&gt; print bedtool.head()

chr7    139371278   139371296   Peak_0  1000    +   5.09062636514014    48.9834262537309    -1  -1
chr7    139371257   139371278   Peak_1  1000    +   5.0840236303159 48.7463712698062    -1  -1
chr7    155781335   155781431   Peak_2  1000    +   3.70481328524336    42.6519289009201    -1  -1
chr7    87156569    87156676    Peak_3  1000    +   3.95023151551588    37.7848384917051    -1  -1
chr7    105073472   105073521   Peak_4  1000    +   4.14556204062503    34.0756845242392    -1  -1
chr7    128761857   128761952   Peak_5  1000    +   4.02131461357736    33.9350181783027    -1  -1
chr7    121296414   121296454   Peak_6  1000    +   3.50632247892067    30.2512926812531    -1  -1
chr7    139368342   139368352   Peak_7  1000    +   4.41912711395099    29.6666535015756    -1  -1
chr7    87155583    87155635    Peak_8  1000    +   4.08769554637519    29.3752024210392    -1  -1
chr7    105540000   105540028   Peak_9  1000    +   4.2212263105571 29.0451450847765    -1  -1

&gt;&gt;&gt; print type(bed)

&lt;class 'pandas.core.frame.DataFrame'&gt;

&gt;&gt;&gt; print type(bedtool)

&lt;class 'pybedtools.bedtool.BedTool'&gt;
</code></pre>

<hr />
<h2 id="getpeaksexons"><strong><em>GetPeaksExons</em></strong></h2>
<p>Annotates a bedtool, BED narrow peak.</p>
<ul>
<li><strong><code>bed</code></strong> a pandas dataframe in bed format</li>
<li>
<p><strong><code>parsedGTF</code></strong> a parsed GTF file as outputed by parseGTF()</p>
</li>
<li>
<p><strong><code>returns</code></strong> a Pandas dataframe</p>
</li>
</ul>
<pre><code class="python">&gt;&gt;&gt; import AGEpy as age

&gt;&gt;&gt; print bed.head()

chrom chromStart   chromEnd    name score strand       signalValue  \
0  chr7  139371278  139371296  Peak_0  1000      +  5.09062636514014   
1  chr7  139371257  139371278  Peak_1  1000      +   5.0840236303159   
2  chr7  155781335  155781431  Peak_2  1000      +  3.70481328524336   
3  chr7   87156569   87156676  Peak_3  1000      +  3.95023151551588   
4  chr7  105073472  105073521  Peak_4  1000      +  4.14556204062503   

     -log10(pValue) -log10(qvalue) peak  
0  48.9834262537309             -1   -1  
1  48.7463712698062             -1   -1  
2  42.6519289009201             -1   -1  
3  37.7848384917051             -1   -1  
4  34.0756845242392             -1   -1

&gt;&gt;&gt; GTF=age.readGTF(&quot;/beegfs/group_bit/data/projects/departments/Bioinformatics/bit_RNAseq_eCLIP/downloads/gencode.v24.primary_assembly.annotation.gtf&quot;)
&gt;&gt;&gt; print GTF.head()

seqname  source     feature  start    end score strand frame  \
0    chr1  HAVANA        gene  11869  14409     .      +     .   
1    chr1  HAVANA  transcript  11869  14409     .      +     .   
2    chr1  HAVANA        exon  11869  12227     .      +     .   
3    chr1  HAVANA        exon  12613  12721     .      +     .   
4    chr1  HAVANA        exon  13221  14409     .      +     .   

                                         attribute  
0  gene_id &quot;ENSG00000223972.5&quot;; gene_type &quot;transc...&quot;  
1  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
2  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
3  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  
4  gene_id &quot;ENSG00000223972.5&quot;; transcript_id &quot;EN...&quot;  

&gt;&gt;&gt; GTFpa=age.parseGTF(GTF)
&gt;&gt;&gt; print GTFpa.head()

seqname  source     feature  start    end score strand frame gene_status  \
0    chr1  HAVANA        gene  11869  14409     .      +     .       KNOWN   
1    chr1  HAVANA  transcript  11869  14409     .      +     .       KNOWN   
2    chr1  HAVANA        exon  11869  12227     .      +     .       KNOWN   
3    chr1  HAVANA        exon  12613  12721     .      +     .       KNOWN   
4    chr1  HAVANA        exon  13221  14409     .      +     .       KNOWN   

            havana_gene    ...               exon_id      transcript_id  \
0  OTTHUMG00000000961.2    ...                   NaN                NaN   
1  OTTHUMG00000000961.2    ...                   NaN  ENST00000456328.2   
2  OTTHUMG00000000961.2    ...     ENSE00002234944.1  ENST00000456328.2   
3  OTTHUMG00000000961.2    ...     ENSE00003582793.1  ENST00000456328.2   
4  OTTHUMG00000000961.2    ...     ENSE00002312635.1  ENST00000456328.2   

  exon_number  ont     havana_transcript ccdsid transcript_name  \
0         NaN  NaN                   NaN    NaN             NaN   
1         NaN  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   
2           1  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   
3           2  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   
4           3  NaN  OTTHUMT00000362751.1    NaN     DDX11L1-002   

                            gene_type transcript_status gene_name  
0  transcribed_unprocessed_pseudogene               NaN   DDX11L1  
1  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1  
2  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1  
3  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1  
4  transcribed_unprocessed_pseudogene             KNOWN   DDX11L1

&gt;&gt;&gt; bedAn=age.GetPeaksExons(bed,GTFpa)
&gt;&gt;&gt; print bedAn.head()

chrom  chromStart   chromEnd     name  score strand  signalValue  \
0  chr7   155781335  155781431   Peak_2   1000      +     3.704813   
1  chr7   155781335  155781431   Peak_2   1000      +     3.704813   
2  chr7   121296414  121296454   Peak_6   1000      +     3.506322   
3  chr7    87155538   87155583  Peak_16   1000      +     4.077391   
4  chr7   107904733  107904812  Peak_17   1000      +     3.674368   

   -log10(pValue)  -log10(qvalue)  peak         ...           \
0       42.651929              -1    -1         ...            
1       42.651929              -1    -1         ...            
2       30.251293              -1    -1         ...            
3       22.798739              -1    -1         ...            
4       21.118496              -1    -1         ...            

              gene_id  exon_id_count  exon_id norm. mean -log10(pValue)  \
0  ENSG00000184863.10              1                          42.651929   
1  ENSG00000184863.10              1                          42.651929   
2  ENSG00000106034.17              1                          30.251293   
3  ENSG00000135164.18              3                        2951.868281   
4  ENSG00000091140.12              1                          21.118496   

  exon_id signalValue transcript_id_count  \
0            3.704813                   1   
1            3.704813                   1   
2            3.506322                   1   
3           42.703999                   3   
4            3.674368                   1   

  transcript_id norm. mean -log10(pValue)  transcript_id signalValue  \
0                               42.651929                   3.704813   
1                               42.651929                   3.704813   
2                               30.251293                   3.506322   
3                             2951.868281                  42.703999   
4                               21.118496                   3.674368   

  gene_id_count gene_id norm. mean -log10(pValue)  gene_id signalValue  
0             4                        116.619012            17.830941  
1             4                        116.619012            17.830941  
2             2                         30.251293             2.144090  
3             8                       3300.707425            73.902289  
4             5                        135.139064            22.210269  
</code></pre>

<p><strong>gene_id_count</strong>: number of intervals overlapping this gene</p>
<p><strong>transcript_id_count</strong>: number of intervals overlapping this transcript</p>
<p><strong>exon_id_count</strong>: number of intervals overlapping this exon</p>
<hr />
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../biom/" class="btn btn-neutral float-right" title="biom">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../../cookbook/" class="btn btn-neutral" title="Cookbook"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href="../../cookbook/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../biom/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js"></script>
      <script src="../../search/require.js"></script>
      <script src="../../search/search.js"></script>

</body>
</html>
